src/Native/Unix/System.Security.Cryptography.Native.Apple/pal_x509.c

```diff
@@ -52,7 +53,27 @@ AppleCryptoNative_X509GetPublicKey(SecCertificateRef cert, SecKeyRef* pPublicKey
     if (cert == NULL || pPublicKeyOut == NULL || pOSStatusOut == NULL)
         return kErrorBadInput;
 
-    *pOSStatusOut = SecCertificateCopyPublicKey(cert, pPublicKeyOut);
+    // SecCertificateCopyPublicKey was deprecated in 10.14, so use SecCertificateCopyKey on the systems that have it (10.14+),
+    // and SecCertificateCopyPublicKey on the systems that don’t.
+    static SecKeyRef (*secCertificateCopyKey)(SecCertificateRef);
+    static OSStatus (*secCertificateCopyPublicKey)(SecCertificateRef, SecKeyRef*);
+    static int checked;
+    
+    if (!checked)
+    {
+        secCertificateCopyKey = (SecKeyRef (*)(SecCertificateRef))dlsym(RTLD_DEFAULT, "SecCertificateCopyKey");
+        secCertificateCopyPublicKey = (OSStatus (*)(SecCertificateRef, SecKeyRef*))dlsym(RTLD_DEFAULT, "SecCertificateCopyPublicKey");
```

Comment:
@janvorli is there a way we could call dlsym only once in this code snippet?

Reply:
_From @bartonjs on Jul 3, 2018, 8:35 AM PDT_

_In C# I'd do this as `Volatile.Write(ref checked, 1)`, or declare `checked` as `volatile`.  Is declaring `checked` `volatile` here the right way to prevent a reordering race condition, @janvorli?_

https://github.com/maryamariyan/corefx/pull/25#discussion_r199853427

Reply:
_From @bartonjs on Jul 3, 2018, 8:37 AM PDT_


_(Or is my `static local`-fu weak, and we don't need to do a lazy initializer manually, because static locals guarantee that they only run their initializer once? (and therefore the two dlsym calls should move up to variable initializers))_

https://github.com/maryamariyan/corefx/pull/25#discussion_r199854401

Reply:
@jkotas will also know for sure

Reply:
Right, this code is wrong. It can be broken even on x64, depending on the compiler optimizations.

The other places in the corefx shims use volatile, double-checked locking and `pthread_mutex_lock` (look for `pthread_mutex_lock` under src/native for examples). You can use the same pattern.

Reply:
Actually, I would just use `pthread_once` here. It does exactly what you need here.

Reply:
@jkotas, in my case here, I would need to call either of `SecCertificateCopyKey` or `SecCertificateCopyPublicKey` once, and both have arguments and neither returns void.

While looking at the doc [here](http://pubs.opengroup.org/onlinepubs/7908799/xsh/pthread_once.html), it seems that `init_routine()` which is the second argument passed to `pthread_once`, apparently should have no arguments and returns void. 

Please let me know if I misunderstood, but I think because of this observation I don't think I can use pthread_once for my case here.

I just pushed up a commit using the alternative `pthread_mutex_lock` pattern you recommended as seen in other places in corefx.